bin/PEPS.py

```diff
@@ -0,0 +1,109 @@
+        order for now"""
+        # Statetensor
+        sublist = []
+        for i,t in enumerate(self.vertex_tensors):
+            sublist.append(self.vertex_labels[i])
+            sublist.append(t.shape)
+        print(sublist)
+        self.code_statetensor = oe.contract_expression(*sublist)
+        # Inner product
+        sublist = []
+        largest_label = self.virtual_lables.max()
+        for i,t in enumerate(self.vertex_tensors):
+            sublist.append(self.vertex_labels[i])
+            sublist.append(t.shape)            
+            t_c_labels = self.vertex_labels[i] + largest_label
+            t_c_labels[-1] = self.vertex_labels[i][-1]
+            sublist.append(t_c_labels)
+            sublist.append(t.shape)
+        self.code_inner_product = oe.contract_expression(*sublist)
+
+    def copy(self):
+        return PEPS(self.physical_labels,
+                    self.virtual_lables,
+                    self.vertex_labels,
+                    [t.clone() for t in self.vertex_tensors])
+
+    def conj(self):
+        for i in range(self.N):
+            self.vertex_tensors[i] = self.vertex_tensors[i].conj()
+        return self
+
+    def output_variables(self):
+        return torch.cat([t.flatten() for t in self.vertex_tensors])
+
+    def load_variables(self, variables):
+        new_tensors = []
+        index = 0
+        for i, t in enumerate(self.vertex_tensors):
+            num_elements = np.prod(t.shape)
+            new_tensor_vars = variables[index : index + num_elements]
+            index += num_elements
+            new_tensors.append(new_tensor_vars.reshape(t.shape))
+        self.vertex_tensors = new_tensors
+
+    def inner_product(self):
+        largest_label = np.max(self.virtual_lables)
+        sublist = []
+        for i in range(self.N):
+            sublist.append(self.vertex_tensors[i])
+            sublist.append(self.vertex_labels[i])
+            sublist.append(self.vertex_tensors[i].conj())
+            t_c_labels = self.vertex_labels[i].copy() + largest_label
+            t_c_labels[-1] = self.vertex_labels[i][-1]
+            sublist.append(t_c_labels)
+        return oe.contract(*sublist)
+        # expr = self.code_inner_product
+        # inner_product = expr(tensor)
+        # return inner_product
+    
+    def overlap_with(self, peps2):
+        largest_label = np.max(self.virtual_lables)
+        sublist = []
+        for i in range(self.N):
+            sublist.append(self.vertex_tensors[i])
+            sublist.append(self.vertex_labels[i])
+            sublist.append(peps2.vertex_tensors[i].conj())
+            t_c_labels = self.vertex_labels[i].copy() + largest_label
+            t_c_labels[-1] = self.vertex_labels[i][-1]
+            sublist.append(t_c_labels)
+        return oe.contract(*sublist)
+
+    def apply_Hamiltonian(self, H):
+        for h in H:
+            t, i = h
+            self.vertex_tensors[i] = oe.contract('...i,ij->...j', self.vertex_tensors[i], t)
```

Comment:
Physical label is always the last dimension right? For the design perspective, I would suggest using a more explicit convention. Imagine your student wants to read your code.